### DISCUSSION TITLE
No interval in applying buffs causes massive medical skill gain (foods and drinks)

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#14379] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

When applying buffs using a statuseffect with no interval (per frame), it's possible to get massive amounts of skill.

Example:
Vanilla energy drink (applies at 0 condition))
vs
Hungry Europans energy drink (no interval)
vs
Hungry Europans cola (with interval)

https://github.com/user-attachments/assets/d088724b-3ca3-4ebd-bc17-fc204d0e07cf

Modded Energy drink xml
Uses a shootable, meaning you must hold RMB+LMB to drink

```xml
    <Item name="Energy Drink" identifier="he-energydrink1" category="Misc" subcategory="Drinks" maxstacksize="32" maxstacksizecharacterinventory="8" cargocontaineridentifier="metalcrate" description="A refreshing energy drink. Good for when you're absolutely shit-faced and need to sober up a bit." Tags="smallitem,drinkitem" useinhealthinterface="true" scale="0.5" impactsoundtag="impact_aluminiumcan_full" impacttolerance="5.5" isshootable="true" requireaimtouse="true" canflipx="false">
      <PreferredContainer secondary="wreckmedcab,abandonedmedcab" amount="1" spawnprobability="0.01" />
      <PreferredContainer secondary="abandonedcrewcab" minamount="1" maxamount="1" spawnprobability="0.01" />
      <PreferredContainer secondary="crewcab" minamount="0" maxamount="8" spawnprobability="0.1" />
      <PreferredContainer primary="kitchen" secondary="crewcab"/>
      <Fabricate suitablefabricators="vendingmachinedrinks,cafeteria" requiredtime="3" requiredmoney="25" fabricationlimitmin="0" fabricationlimitmax="6" />
      <Deconstruct time="2">
        <Item identifier="aluminium" amount="1" outcondition="0.5"/>
      </Deconstruct>
      <InventoryIcon texture="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Items/Food/DrinksAndIngredients.png" sourcerect="474,436,18,43" origin="0.5,0.5" />
      <Sprite texture="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Items/Food/DrinksAndIngredients.png" sourcerect="474,436,18,43" depth="0.6" origin="0.5,0.5" />
      <ContainedSprite allowedcontainertags="table" texture="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Items/Food/DrinksAndIngredients.png" sourcerect="474,436,18,43" depth="0.525" origin="0.5,0.78"  />
      <BrokenSprite texture="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Items/Food/DrinksAndIngredients.png" sourcerect="493,433,18,46" depth="0.6" maxcondition="99" />
      <Body width="18" height="43" density="11" />
      <Holdable canbepicked="true" canBeCombined="true" removeOnCombined="false" slots="Any,RightHand,LeftHand" aimpos="26,22" holdpos="30,-15" holdangle="0" handle1="0,-5" aimangle="105" aimable="false" swingwhenaiming="true" swingamount="3,9" swingspeed="0.2" msg="ItemMsgPickUpSelect">
        <!-- Regular drinking -->
        <StatusEffect type="OnSecondaryUse" target="This" Condition="99" setvalue="true" comparison="And">
          <Conditional Condition="equals 100" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Drinks/OpenCan.ogg" loop="false" range="200" volume="1.5" />
        </StatusEffect>
        <StatusEffect type="OnUse" target="This" Condition="-12.0" />
        <StatusEffect type="OnUse" target="Character" >
          <ReduceAffliction identifier="drunk" amount="1.5" />
          <ReduceAffliction identifier="drunknodebuffs" amount="1.5" />
          <ReduceAffliction identifier="hallucinating" amount="1.5" />
          <Affliction identifier="haste" amount="12.0" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Drinks/Drinking.ogg" loop="true" range="200" volume="5.0" />
        </StatusEffect>
        <StatusEffect type="OnBroken" target="This">
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush1.ogg" range="350" volume="3" selectionmode="Random" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush2.ogg" range="350" volume="3" selectionmode="Random" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush3.ogg" range="350" volume="3" selectionmode="Random" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush4.ogg" range="350" volume="3" selectionmode="Random" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush5.ogg" range="350" volume="3" selectionmode="Random" />
          <Sound file="%ModDir:2532991202%/Integrations/Hungry Europans/Content/Sounds/Items/Misc/CanCrush6.ogg" range="350" volume="3" selectionmode="Random" />
          <Remove />
          <Spawnitem identifier="he-emptycanenergy1" spawnposition="SameInventory" spawnifinventoryfull="true" />
        </StatusEffect>
      </Holdable>
    </Item>
```

### Reproduction steps

1. Download Dynamic Europa (has the foods and drinks)
2. Drink a can of energy drink

### Bug prevalence

Just once

### Single player or multiplayer?

Single player

### Version

v1.5.9.1 (Summer Update Hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by Regalis11** [comment]
Thank you for the report! This doesn't actually seem to have anything to do with the interval, the issue has to do with the buff being constantly while the item is being used, not once when the item is used.

**Reply by ubertpendragon** [reply]
The interval did fix the massive skill point game though (while applying the same amount of buff once adjusted), so something is afoot.

Why are we doing this that way? Because applying the buff at the end of the item is "uncool". RMB+LMB is the superior way of making drinkeables and eatables.

Compare Hungry Europans energy drink to Vanilla energy drink and see the difference.

**Reply by Regalis11** [reply]
Yeah, there is definitely a bug here, but the issue isn't technically caused by the interval. It seems deltatime wasn't being taken into account in the skill gain calculations, which meant the skill increase was applied as if you had given the character 12 haste per frame, as opposed to 12 haste per second. Using the interval can be used to work around this, because it makes the effect actually give 12 haste per frame (and the appropriate amount of skill), but only does that once per e.g. second. Another way to work around it would be to configure the value as "haste per frame", and set `disabledeltatime="true"`.

The reason why this doesn't affect the vanilla energy drink for example is that it only applies the effect once, not over time, so deltatime doesn't need to be taken into account. I've [created a ticket about this](https://github.com/FakeFishGames/Barotrauma/issues/14404) and we've got a fix implemented in our private repo, coming up in a future update. Sorry for the inconvenience!

**Comment by Rokvach** [comment]
Fixed and tested the fix to be working correctly. Fix coming in the next update. Closing discussion.
